[MODULE1-PART-3] Change an architecture of the project

snikolenko-flo opened the pull request:
Change an architecture of the project. Every module should have a handler that uses manager methods. And manager should use service methods.

snikolenko-flo committed:
refactor: create a handler for validation a user input

create a handler for validation a user input

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a manager for user input validation

create a manager for user input validation

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create email and password validation services for login

create email and password validation services for login

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a service for handling email validation result

create a service for handling email validation result

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a service function for handling password validation result

create a service function for handling password validation result

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: rename a service function that handles email validation result

rename service function that handles email validation result

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put all validation files into a separate folder

put all validation files into a separate folder

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a handler for a submit event

create a handler for a submit event

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a manager for a submit event

create a manager for a submit event

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a service for a submit module

create a service for a submit module

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a handler for rendering a gallery page

create a handler for rendering a gallery page

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a manager for rendering a gallery page

create a manager for rendering a gallery page

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a handler for re rendering a gallery page

create a handler for re rendering a gallery page

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a handler for redirecting to a login page

create a handler for redirecting to a login page if a token does not exist

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a handler for setting a token expire time

create a handler for setting a token expire time after a page reloading

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a service for a clicked page

create a service function that returns a number of a clicked page

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put the fetch function into a gallery manager

put the fetch function into a gallery manager

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put html wrapping functions into a gallery service

put html wrapping functions into a gallery service

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: combine two handlers in one for login module

combine two handlers in one for login module

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: combine two managers in one for login module

combine two managers in one for login module

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: combine two services in one for login module

combine two services in one for login module

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put validation services to login service

put validation services to login service

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put token checking function to the gallery service

put a function that checks if token exists to the gallery service

MODULE-1-PART-3-gallery

snikolenko-flo committed:
chore: remove redundant spaces

remove redundant spaces

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: rename a function for submitting user data

rename a function for submitting user data

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create function for redirecting user to a gallery page

create function for redirecting user to a gallery page

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor:  change names of validation handler functions

change names of the functions that validates email and password

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: change name of a token expire function

change name of a token expire function

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: move function for getting page number to the gallery manager

move function for getting page number to the gallery manager

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: move a method for adding page number to url to the gallery manager

move a method for adding page number to url to the gallery manager

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: move a method for redirecting to login page to the gallery manager

move a method for redirecting to login page to the gallery manager

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a function for calculating a token expire time

create a function for calculating a token expire time

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: create a function for setting a timer

create a function for setting a timer for token expiring

MODULE-1-PART-3-gallery

snikolenko-flo committed:
docs: add info about project structure

add info about project structure

MODULE-1-PART-3-gallery

khoroshilov-flo commented on gallery/handler.js:
```diff
@@ -0,0 +1,55 @@
+}
+
+export async function reRenderGalleryPage() {
+    return async function(event) {
```
Можно просто экспортировать обработчик события, не оборачивая его в другую функцию

khoroshilov-flo commented on gallery/handler.js:
```diff
@@ -0,0 +1,55 @@
+        if (!clickedPageNumber) return;
+
+        try {
+            const response = await manager.fetchImages(clickedPageNumber);
```
Предлагаю перенести год с response.json() и response.ok в метод fetchImages. Тем самым тебе не нужно будет каждый раз конвертировать ответ в json и делать проверку статуса. Все это можно сделать в fetchImages и вернуть готовый js объект или ошибку если ответ не ok.

khoroshilov-flo commented on gallery/gallery.manager.js:
```diff
@@ -0,0 +1,74 @@
+import { BASE_URL } from "../data/constants.js";
+import { GalleryService } from "./gallery.service.js";
+
+export class GalleryManager {
```
В коде галереи есть код, относящийся к проверке токена. Проверка токена должна быть отделена в отдельный модуль. Каждая страница, которая требует токен для авторизации должна подключать этот модуль. В коде галереи нужно оставить только то, что к ней относится. Это позволяет не смешивать между собой ответственности разных модулей приложения. Один модуль отвечает за авторизацию и проверку токена, другой за запросы к API, третий за работу с картинками.

khoroshilov-flo replied:
Можно еще пойти дальше в ООПе и разделить код галереи по обязанностям. Сразу заметно, что есть часть кода, которая рендерить данные на странице. Есть часть кода, которая запрашивает данные у API. Можно вынести эти под-части и сделать композицию.
```typescript
class RenderGalleryHTML {}
class GalleryAPI {}
class Gallery {
  render = new RenderGalleryHTML();
  api = new GalleryAPI();
}
```
Вот тут затронута эта тема https://ru.wikipedia.org/wiki/%D0%9F%D1%80%D0%B8%D0%BD%D1%86%D0%B8%D0%BF_%D0%B5%D0%B4%D0%B8%D0%BD%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D0%BE%D0%B9_%D0%BE%D1%82%D0%B2%D0%B5%D1%82%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D0%BE%D1%81%D1%82%D0%B8#:~:text=single%2Dresponsibility%20principle%2C%20SRP),%D0%B8%D1%81%D0%BA%D0%BB%D1%8E%D1%87%D0%B8%D1%82%D0%B5%D0%BB%D1%8C%D0%BD%D0%BE%20%D0%BD%D0%B0%20%D0%BE%D0%B1%D0%B5%D1%81%D0%BF%D0%B5%D1%87%D0%B5%D0%BD%D0%B8%D0%B5%20%D1%8D%D1%82%D0%BE%D0%B9%20%D0%BE%D1%82%D0%B2%D0%B5%D1%82%D1%81%D1%82%D0%B2%D0%B5%D0%BD%D0%BD%D0%BE%D1%81%D1%82%D0%B8.

khoroshilov-flo commented on services/url-manipulation.service.js:
```diff
@@ -0,0 +1,16 @@
+export class UrlManipulationService {
+    getPageNumberFromUrl() {
+        const currentUrl = window.location.search;
+        const pageNumber = Number(currentUrl.split('?page=')[1]);
```
Вот с этим будет удобнее https://developer.mozilla.org/en-US/docs/Web/API/URLSearchParams

khoroshilov-flo replied:
Также стоит добавить больше проверок значения и улучшить парсинг его. В этом помогут isFinite, isNaN, parseInt.

khoroshilov-flo commented on services/url-manipulation.service.js:
```diff
@@ -0,0 +1,16 @@
+    }
+
+    redirectToPage(pageNumber) {
+        window.location.href = `../gallery/gallery.html?page=${pageNumber}`;
```
Это стоит убрать в модуль галереи, т.к. этот сервис от него отделен(абстрактен)

snikolenko-flo committed:
refactor: remove function wrapping for event handlers

remove function wrapping for event handlers

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put response.json and response.ok to the fetchImages method

put response.json and response.ok to the fetchImages method

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put a token methods to a separate module

put a token methods to a separate module

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: use the search params interface

use the search params interface instead of the split method

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: add more checks for a page number

add more checks for a page number

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: move page redirecting to a gallery service

move page redirecting to a gallery service

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: split gallery module to render and api parts

split gallery module to render and api parts

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: put response.json and response.ok to the get token method

put response.json and response.ok to the get token method

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: make a password validation function more readable

make a password validation function more readable

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: change names of the functions in the gallery module

change names of the functions that fetch and load images in the gallery module

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: add a separate function for to save a token

add a separate function for to save a token

MODULE-1-PART-3-gallery

snikolenko-flo committed:
refactor: separate a token manager to several classes

separate a token manager to several classes

MODULE-1-PART-3-gallery

khoroshilov-flo commented:
Можно мержить и двигаться к след. части модуля, все окей.
